### PR TITLE
Use frustum culling in OSM tile traversal

### PR DESCRIPTION
**Pull request by Pessimistress**
For #4482 

#### Change List
- Use frustum culling in OSM tile traversal
- Show tile boundary in website demo

### CODE REVIEW

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    // Avoid loading overlapping tiles - if a descendant is requested, do not request the ancester
+    if (!this.childVisible) {
+      let {z} = this;
+      if (z < maxZ && z >= minZ) {
+        // Adjust LOD by distance to camera
+        const distance =
+          (boundingVolume.distanceTo(viewport.cameraPosition) * viewport.scale) / viewport.height;
+        z += Math.floor(Math.log2(distance));
```
Comment (xintongxia) [inline modules/geo-layers/src/tile-layer/tile-2d-traversal.js:46]: why do you need apply `* viewport.scale / viewport.height` to  `(boundingVolume.distanceTo(viewport.cameraPosition)`? Maybe add a comment about your calculation?

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  const minZ = viewport.pitch <= 60 ? maxZ : 0;
+
+  const root = new OSMNode(0, 0, 0);
+  const traversalParams = {
+    viewport,
+    cullingVolume,
+    elevationBounds: [elevationMin, elevationMax],
+    minZ,
+    maxZ,
+    offset: 0
```
Comment (xintongxia) [inline modules/geo-layers/src/tile-layer/tile-2d-traversal.js:112]: what is `offset`?

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  );
+  const cullingVolume = new CullingVolume(planes);
+
+  // Project zRange from meters to common space
+  const unitsPerMeter = viewport.distanceScales.unitsPerMeter[2];
+  const elevationMin = (zRange && zRange[0] * unitsPerMeter) || 0;
+  const elevationMax = (zRange && zRange[1] * unitsPerMeter) || 0;
+
+  // Always load at the current zoom level if pitch is small
+  const minZ = viewport.pitch <= 60 ? maxZ : 0;
```
Comment (kylebarron) [inline modules/geo-layers/src/tile-layer/tile-2d-traversal.js:103]: Is this necessary to keep in sync with basemaps, e.g. Mapbox GL? I could envision cases where a user might want to relax this to load fewer tiles at moderate pitch, e.g. 40-60 degrees.

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  // Check worlds in repeated maps
+  traversalParams.offset = -1;
+  while (root.update(traversalParams)) {
+    if (--traversalParams.offset < -MAX_MAPS) {
+      break;
+    }
+  }
+  traversalParams.offset = 1;
+  while (root.update(traversalParams)) {
+    if (++traversalParams.offset > -MAX_MAPS) {
```
Comment (kylebarron) [inline modules/geo-layers/src/tile-layer/tile-2d-traversal.js:126]: Should this be positive `MAX_MAPS`?

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import {CullingVolume, Plane, AxisAlignedBoundingBox} from '@math.gl/culling';
+
+const TILE_SIZE = 512;
+// number of world copies to check
+const MAX_MAPS = 3;
```
Comment (kylebarron) [inline modules/geo-layers/src/tile-layer/tile-2d-traversal.js:5]: Should checking world copies be configurable? If I turn off `repeat: true`

image

The `onViewportLoad` callback in this viewport still holds 75 tiles, while I only count 15 tiles in the viewport. Anecdotally, it seems that many more tiles are loaded when looking west than when looking north.

File: examples/website/map-tile/app.js [hunk 6]
```diff
@@ -49,8 +50,8 @@ export default class App extends PureComponent {
 
         pickable: true,
         onHover: this._onHover,
-        autoHighlight,
-        highlightColor,
+        autoHighlight: showBorder,
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply examples/website/map-tile/app.js:53]: This is an example.

File: modules/geo-layers/src/tile-layer/tile-2d-traversal.js (hunk 3 above)
Comment (Pessimistress) [reply modules/geo-layers/src/tile-layer/tile-2d-traversal.js:105]: This is also what mapbox-gl is doing. Lowering the threshold to 40-50 saves 2-3 tile loads tops, and may have unexpected behavioral changes for existing users (60 is the default `maxPitch`)

**Pull request merged**